integration-tests/src/tests/client/features/nonrefundable_transfer.rs

```diff
@@ -0,0 +1,346 @@
+
+    // Delete the new account (that has 1 NEAR of non-refundable balance).
+    let delete_account_tx_result = delete_account(&mut env, &new_account);
+    delete_account_tx_result.unwrap().assert_success();
+    assert!(!account_exists(&mut env, new_account_id));
+}
+
+/// Non-refundable balance cannot be transferred.
+#[test]
+fn non_refundable_balance_cannot_be_transferred() {
+    let mut env = setup_env();
+    let new_account_id: AccountId = "subaccount.test0".parse().unwrap();
+    let new_account = InMemorySigner::from_seed(
+        new_account_id.clone(),
+        KeyType::ED25519,
+        new_account_id.as_str(),
+    );
+    // The `new_account` is created with 1 NEAR non-refundable balance.
+    let create_account_tx_result = exec_transfer(
+        &mut env,
+        signer(),
+        new_account_id.clone(),
+        NEAR_BASE,
+        true,
+        true,
+        false,
+        false,
+    );
+    create_account_tx_result.unwrap().assert_success();
+
+    // Although `new_account` has 1 NEAR non-refundable balance, it cannot make neither refundable nor non-refundable transfer of 1 yoctoNEAR.
+    for nonrefundable in [false, true] {
+        let transfer_tx_result = exec_transfer(
+            &mut env,
+            new_account.clone(),
+            receiver(),
+            1,
+            nonrefundable,
+            false,
+            false,
+            false,
+        );
+        match transfer_tx_result {
+            Err(InvalidTxError::NotEnoughBalance { signer_id, balance, .. }) => {
+                assert_eq!(signer_id, new_account_id);
+                assert_eq!(balance, 0);
+            }
+            _ => panic!("Expected NotEnoughBalance error"),
+        }
+    }
+}
+
+/// Non-refundable balance allows to have account with zero balance and more than 1kB of state.
+#[test]
+fn non_refundable_balance_allows_1kb_state_with_zero_balance() {
+    let mut env = setup_env();
+    let new_account_id: AccountId = "subaccount.test0".parse().unwrap();
+    let tx_result =
+        exec_transfer(&mut env, signer(), new_account_id, NEAR_BASE / 5, true, true, false, true);
+    tx_result.unwrap().assert_success();
+}
+
+/// Non-refundable transfer successfully adds non-refundable balance when creating named account.
+#[test]
+fn non_refundable_transfer_create_named_account() {
+    let new_account_id: AccountId = "subaccount.test0".parse().unwrap();
+    let tx_result =
+        exec_transfer(&mut setup_env(), signer(), new_account_id, 1, true, true, false, false);
+    tx_result.unwrap().assert_success();
+}
+
+/// Non-refundable transfer successfully adds non-refundable balance when creating NEAR-implicit account.
+#[test]
+fn non_refundable_transfer_create_near_implicit_account() {
+    let new_account_id = near_implicit_test_account();
+    let tx_result =
+        exec_transfer(&mut setup_env(), signer(), new_account_id, 1, true, true, true, false);
+    tx_result.unwrap().assert_success();
```

Comment:
I think there should be a view call here to check the implicit account really does have the non-refundable balance it is supposed to have.

Reply:
It is checked here:
https://github.com/near/nearcore/pull/9600/files#diff-14b73a0d63728329a6f04a38d1d2a7423bdd89edf2095debc5e94a8e48045f29R173